File: solutions/ruby/clock/1/clock.rb
class Clock
  TIME_FORMAT = '%02<hour>u:%02<minute>u'
  
  attr_accessor :hour
  attr_accessor :minute

  def initialize(hour: 0, minute: 0)
    self.minute = minute % 60
    self.hour = (hour + (minute - self.minute) / 60) % 24
  end

  def to_s
    TIME_FORMAT % {hour: self.hour, minute: self.minute}
  end

  def +(other)
    Clock.new(hour: self.hour + other.hour, minute: self.minute + other.minute)
  end

  def -(other)
    Clock.new(hour: self.hour - other.hour, minute: self.minute - other.minute)
  end

  def ==(other)
    self.hour == other.hour and self.minute == other.minute
  end

end

File: solutions/cpp/clock/1/clock.cpp
#include "clock.h"

namespace date_independent {
    clock::clock(int hour, int minute) {
        hour += minute / 60;
        minute %= 60;
        if (minute < 0) {
            minute += 60;
            hour--;
        }
        hour %= 24;
        if (hour < 0) {
            hour += 24;
        }
        this->hour = hour;
        this->minute = minute;
    }
    clock clock::at(int h, int m) {
        return clock(h, m);
    }
    clock clock::plus(int mins) {
        return clock(this->hour, this->minute + mins);
    }
    clock clock::minus(int mins) {
        return clock(this->hour, this->minute - mins);
    }/*
    clock clock::plus(clock other) {
        return clock(this->hour + other.hour, this->minute + other.minute);
    }
    clock clock::minus(clock other) {
        return clock(this->hour - other.hour, this->minute - other.minute);
    }*/
    clock::operator std::string() const {
        auto str_hour =
            this->hour < 10 ? "0" + std::to_string(this->hour)
                            : std::to_string(this->hour);
        auto str_minute =
            this->minute < 10 ? "0" + std::to_string(this->minute)
                              : std::to_string(this->minute);
        return str_hour + ":" + str_minute;
    }
    bool clock::operator==(clock other) const {
        return this->hour == other.hour && this->minute == other.minute;
    }
    bool clock::operator!=(clock other) const {
        return !(*this == other);
    }
}  // namespace date_independent
